File: src/main/java/random/telegramhomebot/controllers/HostTimeLogController.java
package random.telegramhomebot.controllers;

import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import random.telegramhomebot.model.HostState;
import random.telegramhomebot.model.HostTimeLog;
import random.telegramhomebot.model.TimeLogDto;
import random.telegramhomebot.repository.HostTimeLogRepository;
import random.telegramhomebot.services.TimeLogConverter;

import javax.annotation.Resource;
import java.sql.Timestamp;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.mapping;
import static java.util.stream.Collectors.toList;
import static random.telegramhomebot.AppConstants.Hosts.HOSTS_MAPPING;
import static random.telegramhomebot.AppConstants.HostsTimeLog.END_DATE_REQ_PARAM;
import static random.telegramhomebot.AppConstants.HostsTimeLog.START_DATE_REQ_PARAM;
import static random.telegramhomebot.AppConstants.HostsTimeLog.TIME_LOG_MAPPING;
import static random.telegramhomebot.AppConstants.HostsTimeLog.TIME_LOG_MAP_MODEL_ATTR;
import static random.telegramhomebot.AppConstants.HostsTimeLog.TIME_LOG_VIEW;

@Controller
@RequestMapping(HOSTS_MAPPING)
public class HostTimeLogController {

	@Resource
	private HostTimeLogRepository hostTimeLogRepository;
	@Resource
	private TimeLogConverter converter;

	@RequestMapping(TIME_LOG_MAPPING)
	public String getTimeLogForPeriod(
			@RequestParam(value = START_DATE_REQ_PARAM, defaultValue = "#{T(java.time.LocalDateTime).now()}")
			@DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
			@RequestParam(value = END_DATE_REQ_PARAM, defaultValue = "#{T(java.time.LocalDateTime).now()}")
			@DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate,
			Model model) {
		LocalDateTime startOfDate = startDate.atTime(LocalTime.MIN);
		LocalDateTime endOfDate = endDate.atTime(LocalTime.MAX);

		List<HostTimeLog> logs = hostTimeLogRepository
				.findByCreatedDateBetween(Timestamp.valueOf(startOfDate), Timestamp.valueOf(endOfDate));

		Function<HostTimeLog, String> hostGroupingBy = log -> log.getHost().getDeviceName() != null
				? log.getHost().getDeviceName()
				: log.getId().toString();

		Map<String, List<TimeLogDto>> timeLogMap = logs.stream()
				.peek(log -> {
					if (!(log.getState() == HostState.FAILED)) {
						log.setState(HostState.REACHABLE);
					}
				})
				.collect(groupingBy(hostGroupingBy, mapping(log -> converter.convert(log), toList())));

		model.addAttribute(TIME_LOG_MAP_MODEL_ATTR, timeLogMap);
		return TIME_LOG_VIEW;
	}

}

File: src/main/java/random/telegramhomebot/controllers/HostTimeLogController.kt
package random.telegramhomebot.controllers

import org.springframework.format.annotation.DateTimeFormat
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import random.telegramhomebot.const.AppConstants.DEFAULT_DATE_VALUE_NOW
import random.telegramhomebot.const.AppConstants.END_DATE_REQ_PARAM
import random.telegramhomebot.const.AppConstants.HOSTS_MAPPING
import random.telegramhomebot.const.AppConstants.START_DATE_REQ_PARAM
import random.telegramhomebot.const.AppConstants.TIME_LOG_MAPPING
import random.telegramhomebot.const.AppConstants.TIME_LOG_VIEW
import random.telegramhomebot.db.dto.TimeLogDto
import random.telegramhomebot.db.model.HostTimeLog
import random.telegramhomebot.db.repository.HostTimeLogRepository
import random.telegramhomebot.services.hosts.TimeLogConverter
import random.telegramhomebot.utils.logger
import java.time.LocalDate

@Controller
@RequestMapping(HOSTS_MAPPING)
class HostTimeLogController(
    private val hostTimeLogRepository: HostTimeLogRepository,
    private val timeLogConverter: TimeLogConverter
) {
    val log = logger()

    @RequestMapping(TIME_LOG_MAPPING)
    fun getTimeLogForPeriod(
        @RequestParam(value = START_DATE_REQ_PARAM, defaultValue = DEFAULT_DATE_VALUE_NOW)
        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) startDate: LocalDate,
        @RequestParam(value = END_DATE_REQ_PARAM, defaultValue = DEFAULT_DATE_VALUE_NOW)
        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) endDate: LocalDate,
        model: Model
    ): String {
//        val startOfDate = startDate.atTime(LocalTime.MIN)
//        val endOfDate = endDate.atTime(LocalTime.MAX)
//        val logs = hostTimeLogRepository.findByCreatedDateBetween(valueOf(startOfDate), valueOf(endOfDate))
//        model.addAttribute(TIME_LOG_MAP_MODEL_ATTR, getTimeLogDtoMap(logs))
//        model.addAttribute(START_DATE_MODEL_ATTR, startDate)
//        model.addAttribute(END_DATE_MODEL_ATTR, endDate)
        return TIME_LOG_VIEW
    }

    private fun getTimeLogDtoMap(logs: List<HostTimeLog>): Map<String, List<TimeLogDto>> =
        logs.groupBy({ it.host.deviceName ?: it.host.mac ?: "" }, { timeLogConverter.convertToDto(it) }).toSortedMap()
}
